lib/vm.py

```diff
@@ -72,7 +72,7 @@ def reboot(self, force=False, verify=False):
         return ret
 
     def try_get_and_store_ip(self):
-        ip = self.param_get('networks', '0/ip', accept_unknown_key=True)
+        ip = cast(str, self.param_get('networks', '0/ip', accept_unknown_key=True))
```

Comment:
If this ad-hoc conversion-to-bool will force to `cast(str)` every call to `param_get` that just is not a bool, that seems to highlight that the ad-hoc conversion was a bad idea in the first time.  It would look much wiser to provide a `xe_parse_bool_param()` helper that would be explicitly called exactly where needed, let everyone have a strict typing.

Reply:
I agree, but I don't know where those calls to `param_get` expecting a `bool` are. I'm not even sure some call may not expect one or the other return type depending on the parameter…

I only had to add two `cast`:

~~~
rg cast
lib/vm.py
5:from typing import List, Literal, Optional, Union, cast, overload
75:        ip = cast(str, self.param_get('networks', '0/ip', accept_unknown_key=True))

lib/pool.py
3:from typing import Any, Dict, Optional, cast
33:        self.uuid = cast(str, self.master.xe('pool-list', minimal=True))
~~~

I can add a `param_get_str()` if you find it more convenient, but it won't help with the second cast

Reply:
Or I can rename the existing `param_get()` method in `param_get_any()` (or `param_get_str_or_bool()`) everywhere, and create a new `param_get()` that returns a `Optional[str]`.
I think I like the `cast()` version more though.

Reply:
The problem is the DWIM behavior on bools is in `xe()`, which clearly seems a bad place for this.  Do we even have a clue if it is only `get_param` using this?

> I think I like the cast() version more, though.

You only had to add a single one because of how few type hints we have today.  When the majority of the calls would have to get an explicit cast, IMO it can only be a wrong solution.

And anyway, how are things like this behaving with this hidden cast to bool?

```
lib/vdi.py:        return self.param_get("read-only") == "true"
```

This implicit cast only created a mess.

Reply:
I fully agree. But given the required knowledge of the code base to ensure it works properly, I would rather keep that change for another PR.

Reply:
OK, moved to #291